### PR TITLE
Feature/create a business component

### PR DESCRIPTION
**Pull request by adrien-louis-r**
I am not very confident about this, so I will look forward to your feedback!

### CODE REVIEW

File: docs/create-a-business-component.md [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+## What are we going to build
+
+To explain the concept and the emergeance of modules, we will now add a **store
+locator** to our home page and see how to extract it properly as a module.
+
+Before diving into the code, we have to introduce a new concept: **Enhancers**.
+In this case, it will be responsible to fetch the data from our **GraphQL
+schema**, transform the data and feed it to our component.
+
+The **Enhancers** are not means to only be _data fetchers_, they contain most of
```
Comment (hihuz) [inline docs/create-a-business-component.md]: typo : mean**t**

File: docs/create-a-business-component.md [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+> Unlike UI components, Business ones are often _smart_ and contain logic. We
+> try to extract this logic in **Enhancers**, more on that later.
+
+## What are we going to build
+
+To explain the concept and the emergeance of modules, we will now add a **store
+locator** to our home page and see how to extract it properly as a module.
+
+Before diving into the code, we have to introduce a new concept: **Enhancers**.
+In this case, it will be responsible to fetch the data from our **GraphQL
```
Comment (hihuz) [inline docs/create-a-business-component.md]: I think I would rephrase this to be a bit more clear : "In this specific case, the Enhancer will be responsible for fetching ..."

The first part is a bit subjective but it sounds better to me. 
The second part "responsible for fetching" I think is more correct.

File: docs/create-a-business-component.md [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+schema**, transform the data and feed it to our component.
+
+The **Enhancers** are not means to only be _data fetchers_, they contain most of
+the **Business logic** of our application, we use the
+[Higher-Order Components pattern (HOC)](https://reactjs.org/docs/higher-order-components.html)
+to create them.
+
+> In Front-Commerce, we use a react library:
+> [Recompose](https://github.com/acdlite/recompose) to handle composition of
+> **HOC**, it provides a lots of helpers which are really usefull to _enrich_
```
Comment (hihuz) [inline docs/create-a-business-component.md]: a couple typos :

a lot~~s~~ of helpers 

useful~~l~~

File: docs/create-a-business-component.md [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+> our components.
+
+In the following steps, we are going to build our store locator. It will be a
+simple marker on a map indicating where to find our store.
+
+### Installing the dependencies
+
+To create the map, we are going to use the `react-leaflet` package. It is a
+simple component which will allow us to display OpenStreetMap easely.
```
Comment (hihuz) [inline docs/create-a-business-component.md]: typo : eas**i**ly

File: docs/create-a-business-component.md [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+
+This is one of the biggest advantage of using React to build our front-end, we
```
Comment (hihuz) [inline docs/create-a-business-component.md]: Nit : one of the biggest advantage~~s~~ 

I think plural form is more correct since the phrasing implies that there are multiple advantages

File: docs/create-a-business-component.md [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+
+### Our new Homepage
+
+You don't need to anticipate every **UI** or **Business** components in your
+application. When your component will begin to get bigger, the need to extract
+some of them will come naturally.
+
+To illustrate this point, we are going to first add a hardcoded version of our
+map into the homepage. Then we will extract it into its own module and finally
+we are going to fetch the required data from the **GraphQl schema**.
```
Comment (hihuz) [inline docs/create-a-business-component.md]: Nit : GraphQ**L**

File: docs/create-a-business-component.md [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+          <span>My awesome store is HERE!</span>
+        </Popup>
+      </Marker>
+    </Map>
+  </div>
+);
+// ...
+```
+
+With that, you should see the map appears in your homepage.
```
Comment (hihuz) [inline docs/create-a-business-component.md]: typo : appear~~s~~

File: docs/create-a-business-component.md [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+
+### Extracting our new component
+
+Having the map in the Home component could be fine for a time, but we don't want
+to maintain big bloated components, so we are going to extract it to its own
+Module.
+
+For it is exactly the same as before, we juste moved it to the
```
Comment (hihuz) [inline docs/create-a-business-component.md]: typo : just~~e~~

File: docs/create-a-business-component.md [hunk 9]
```diff
@@ -0,0 +1,345 @@
...
+export default StoreLocatorQuery =>
+  graphql(StoreLocatorQuery, {
+    props: ({ data }) => ({
+      loading: data.loading,
+      store: data.loading ? null : data.store
+    })
+  });
+```
+
+Here, we are using the `graphql` function which allow us to fetch data in our
```
Comment (hihuz) [inline docs/create-a-business-component.md]: typo : allow**s**

File: docs/create-a-business-component.md [hunk 10]
```diff
@@ -0,0 +1,345 @@
...
+graphQL schema. It have two arguments, the first one is the **Query** we need to
```
Comment (hihuz) [inline docs/create-a-business-component.md]: typo : It ha**s**

Although I would have probably said "It takes"

File: docs/create-a-business-component.md [hunk 11]
```diff
@@ -0,0 +1,345 @@
...
+      latitude
+    }
+  }
+}
+```
+
+> You may think that some of those data are already fetched in our EnhanceHome
+> and thus, this is ineficient. But react-apollo will handle that for you and
+> will group the requests to make only the necessary requests.  
+> This allow to only think about what a component needs and responsibility
```
Comment (hihuz) [inline docs/create-a-business-component.md]: I would put it like this : "This allows us to ..."

File: docs/create-a-business-component.md [hunk 12]
```diff
@@ -0,0 +1,345 @@
...
+> for the recovery of its data lies with him.
```
Comment (hihuz) [inline docs/create-a-business-component.md]: recovery seems correct here but it sounds a bit odd to me, what about "retrieval" ?
May need input from someone else before doing this change :slightly_smiling_face: 

also at the end : "lies with it" seems a bit more correct

File: docs/create-a-business-component.md [hunk 13]
```diff
@@ -0,0 +1,345 @@
...
+The major change here is the use of the data from the available props.
+
+When dealing with asynchronous resources like fetching data from the backend,
+you have to handle the loading state. Here we just show a simple `Loading..` to
+the user, but in a real life application, you would want to show placeholders or
+spinners.
+
+> You should also handle error cases, but it is not the purpose of this guide.
+> Just know that you could use
+> [Error Boundaries](https://reactjs.org/blog/2017/07/26/error-handling-in-react-16.html)
```
Comment (hihuz) [inline docs/create-a-business-component.md]: :+1:

File: docs/create-a-business-component.md [hunk 14]
```diff
@@ -0,0 +1,345 @@
...
+      latitude: PropTypes.number.isRequired
+    }).isRequired
+  })
+};
+
+export default EnhanceStoreLocator(StoreLocatoreQuery)(StoreLocator);
+```
+
+> A side note on **Protypes**. The Proptypes allow us to validate the props
+> passed to a child component, this will avoid breaking your app passing invalid
```
Comment (hihuz) [inline docs/create-a-business-component.md]: proposal : "breaking your app **by** passing"

File: docs/create-a-business-component.md [hunk 15]
```diff
@@ -0,0 +1,345 @@
...
+> props.  
+This will also serve as a props documentation for the other developers
+> in your team. They may be a little verbose, but it is important to maintain
+> them updated.
```
Comment (hihuz) [inline docs/create-a-business-component.md]: I think "maintain them updated" is a bit redundant.

I would say : "to keep them updated" or simply "to maintain them"

File: docs/create-a-business-component.md [hunk 16]
```diff
@@ -0,0 +1,345 @@
...
+> If you feel the need to understand why we went for this organization, feel
+> free to refer to the [components history page](#TODO-lien-history) first.
+
+In this section, we will build a **Business Component**. If you have already
+gone through the [Create an UI component](#TODO-lien-create-ui-component), the
+core concept is the same.
+
+## What is a Business component
+
+A Business component will reside in the `src/theme/modules` directory. Those
```
Comment (real34) [inline docs/create-a-business-component.md]: Nit: I am not sure it is « will reside » is correct in this context.

Proposal: « should live »

File: docs/create-a-business-component.md [hunk 17]
```diff
@@ -0,0 +1,345 @@
...
+Quickly, you will want to **extract** some components from your page to avoid a
+_big bloated file_. Some of those components will be extracted as **reusable UI
+components** but some are just very specific to your page and there is no much
+sense to put them in the ui folder.
+
+> They are often a mix between UI components and custom layout. They may be
+> splitted in multiples components if they are big enough.
+
+We put those components in the `src/theme/modules` directory. Generally, they
+are linked to **your business** and often needs backend data like CMS content or
```
[the remark this answers is not in the document]
Comment (real34) [reply docs/create-a-business-component.md]: Nit: I would use « related » over « linked » here

Nit2: « such as » over « like »

File: docs/create-a-business-component.md [hunk 18]
```diff
@@ -0,0 +1,345 @@
...
+
+Quickly, you will want to **extract** some components from your page to avoid a
+_big bloated file_. Some of those components will be extracted as **reusable UI
+components** but some are just very specific to your page and there is no much
+sense to put them in the ui folder.
+
+> They are often a mix between UI components and custom layout. They may be
+> splitted in multiples components if they are big enough.
+
+We put those components in the `src/theme/modules` directory. Generally, they
```
Comment (real34) [inline docs/create-a-business-component.md]: This first sentence may be removed since it is already explained a few paragraphs before IMO

File: docs/create-a-business-component.md [hunk 19]
```diff
@@ -0,0 +1,345 @@
...
+are linked to **your business** and often needs backend data like CMS content or
+store information. We refer to them as **Business components** or even
+**modules**.
+
+> Unlike UI components, Business ones are often _smart_ and contain logic. We
+> try to extract this logic in **Enhancers**, more on that later.
+
+## What are we going to build
+
+To explain the concept and the emergeance of modules, we will now add a **store
```
Comment (real34) [inline docs/create-a-business-component.md]: Typo: « emerge**~a~**nce »

Maybe replacing it with « the need behind modules » or something like that could be more meaningful… but I’m not confident with this one :)

File: docs/create-a-business-component.md [hunk 20]
```diff
@@ -0,0 +1,345 @@
...
+> [Recompose](https://github.com/acdlite/recompose) to handle composition of
+> **HOC**, it provides a lots of helpers which are really usefull to _enrich_
+> our components.
+
+In the following steps, we are going to build our store locator. It will be a
+simple marker on a map indicating where to find our store.
+
+### Installing the dependencies
+
+To create the map, we are going to use the `react-leaflet` package. It is a
```
Comment (real34) [inline docs/create-a-business-component.md]: Nit: linking to https://react-leaflet.js.org/en/ could be nice

File: docs/create-a-business-component.md (hunk 8 above)
Comment (real34) [reply docs/create-a-business-component.md]: I think a word (« For **now** » ?) is missing in this sentence.

File: docs/create-a-business-component.md [hunk 21]
```diff
@@ -0,0 +1,345 @@
...
+Here, we are using the `graphql` function which allow us to fetch data in our
+graphQL schema. It have two arguments, the first one is the **Query** we need to
+fetch (we will handle this part soon), in the second, we use the `props`
+property in the second parameter to compute the **Query** result `data` to a
+`loading` and a `store` properties.
+
+You can find all the available `graphql` options in the
+[**React Apollo** documentation](https://www.apollographql.com/docs/react/api/react-apollo.html)
+
+As you can see, our Enhancer need a **Query**, this a `.gql` (or `.graphql`)
```
Comment (real34) [inline docs/create-a-business-component.md]: typo: need**s**

File: docs/create-a-business-component.md [hunk 22]
```diff
@@ -0,0 +1,345 @@
...
+    }
+    coordinates {
+      longitude
+      latitude
+    }
+  }
+}
+```
+
+> You may think that some of those data are already fetched in our EnhanceHome
```
Comment (real34) [inline docs/create-a-business-component.md]: EnhanceHome needs to be quoted with backticks

File: docs/create-a-business-component.md [hunk 23]
```diff
@@ -0,0 +1,345 @@
...
+> and thus, this is ineficient. But react-apollo will handle that for you and
```
Comment (real34) [inline docs/create-a-business-component.md]: Typo: « inef**f**icient »

File: docs/create-a-business-component.md (hunk 12 above)
Comment (real34) [reply docs/create-a-business-component.md]: « fetching » could be used also.

File: docs/create-a-business-component.md [hunk 24]
```diff
@@ -0,0 +1,345 @@
...
+> will group the requests to make only the necessary requests.  
+> This allow to only think about what a component needs and responsibility
+> for the recovery of its data lies with him.
+
+### Making it dynamic
+
+Now that we have an Enhancer ready, we are going to use it in our StoreLocator.
```
Comment (real34) [inline docs/create-a-business-component.md]: if you mean the `StoreLocator` component, it should be in code tags IMO. Otherwise we may just use « store locator » in 2 words here.

File: docs/create-a-business-component.md [hunk 25]
```diff
@@ -0,0 +1,345 @@
...
+you have to handle the loading state. Here we just show a simple `Loading..` to
+the user, but in a real life application, you would want to show placeholders or
+spinners.
+
+> You should also handle error cases, but it is not the purpose of this guide.
+> Just know that you could use
+> [Error Boundaries](https://reactjs.org/blog/2017/07/26/error-handling-in-react-16.html)
+> for that.
+
+```js
```
Comment (real34) [inline docs/create-a-business-component.md]: If you have the time here, you could try to use diffs to illustrate the changes in a better way.

See https://gist.github.com/salmedina/ad8bea4f46de97ea132f71b0bca73663#file-markdowndiffexample-md for instance

File: docs/create-a-business-component.md [hunk 26]
```diff
@@ -0,0 +1,345 @@
...
+      longitude: PropTypes.number.isRequired,
+      latitude: PropTypes.number.isRequired
+    }).isRequired
+  })
+};
+
+export default EnhanceStoreLocator(StoreLocatoreQuery)(StoreLocator);
+```
+
+> A side note on **Protypes**. The Proptypes allow us to validate the props
```
Comment (real34) [inline docs/create-a-business-component.md]: Typo: « Pro**pT**ypes »
Typo2: « allow**s** »

File: docs/create-a-business-component.md [hunk 27]
```diff
@@ -0,0 +1,345 @@
...
+> passed to a child component, this will avoid breaking your app passing invalid
+> props.  
+This will also serve as a props documentation for the other developers
```
Comment (real34) [inline docs/create-a-business-component.md]: There is a missing `>` on this line

File: docs/create-a-business-component.md [hunk 28]
```diff
@@ -0,0 +1,345 @@
...
+  <div className="page page--home">
+    //...
+    <StoreLocator />
+  </div>
+);
+```
+
+> The store locator we just created is very simple and it has a very limited
+> Business impact. It was for the sake of this guide. A component such as this
+> one might have its place in the ui directory in your application.
```
Comment (real34) [inline docs/create-a-business-component.md]: I think the last sentence may be misleading, since in the beginning we explained why it should live in the `modules` directory.

We could remove it, and only emphasize on the fact that it is a simple and naive implementation to illustrate how to create such components. Splitting code accordingly needs practice and refinement, thus one team could split code differently.

If we want to explain what should go to UI, then there needs to be a clear note about not moving GraphQL code there or things related to the store. But I feel like it would be too much information here.

File: docs/create-a-business-component.md (hunk 8 above)
Comment (hihuz) [reply docs/create-a-business-component.md]: I read it as "Since it is..." or "Because it is..." but maybe the author could clarify :slightly_smiling_face:

File: docs/create-a-business-component.md (hunk 12 above)
Comment (hihuz) [reply docs/create-a-business-component.md]: Indeed, but I think he wanted to avoid repetition since he is using the term quite a bit already :slightly_smiling_face:

File: docs/create-a-business-component.md [hunk 29]
```diff
@@ -0,0 +1,345 @@
...
+> This allow to only think about what a component needs and responsibility
+> for the recovery of its data lies with him.
+
+### Making it dynamic
+
+Now that we have an Enhancer ready, we are going to use it in our StoreLocator.
+The major change here is the use of the data from the available props.
+
+When dealing with asynchronous resources like fetching data from the backend,
+you have to handle the loading state. Here we just show a simple `Loading..` to
```
Comment (JulienPradet) [inline docs/create-a-business-component.md]: Here we will show a simple "Loading..." message to the user

File: docs/create-a-business-component.md [hunk 30]
```diff
@@ -0,0 +1,345 @@
...
+the user, but in a real life application, you would want to show placeholders or
+spinners.
```
Comment (JulienPradet) [inline docs/create-a-business-component.md]: It could be a message about reusing UI components too.

File: docs/create-a-business-component.md [hunk 31]
```diff
@@ -0,0 +1,345 @@
...
+
+> You should also handle error cases, but it is not the purpose of this guide.
+> Just know that you could use
```
Comment (JulienPradet) [inline docs/create-a-business-component.md]: However if you want to learn more about it, you could look at Error Boundaries.

File: docs/create-a-business-component.md (hunk 26 above)
Comment (JulienPradet) [inline docs/create-a-business-component.md]: PropTypes on both

File: docs/create-a-business-component.md (hunk 27 above)
Comment (JulienPradet) [inline docs/create-a-business-component.md]: as ~~a props~~ documentation

Comment (JulienPradet) [reply docs/create-a-business-component.md]: for other developers

File: docs/create-a-business-component.md [hunk 32]
```diff
@@ -0,0 +1,345 @@
...
+> A side note on **Protypes**. The Proptypes allow us to validate the props
+> passed to a child component, this will avoid breaking your app passing invalid
+> props.  
+This will also serve as a props documentation for the other developers
+> in your team. They may be a little verbose, but it is important to maintain
+> them updated.
+
+## Using it in our App
+
+We now have all the required pieces, to finish, we just have to import our brand
```
Comment (JulienPradet) [inline docs/create-a-business-component.md]: We now have extracted all the Store Locator logic. Now, we only have to use our brand and shiny module within the homepage.

File: docs/create-a-business-component.md (hunk 28 above)
Comment (JulienPradet) [inline docs/create-a-business-component.md]: I'd add a comment about what to do next.

Either about "we need the map elsewhere, how would we reuse it?" or "we don't have enough data in our schema, where should we find it?" or both :)

File: docs/create-a-business-component.md [hunk 33]
```diff
@@ -0,0 +1,345 @@
...
+graphQL schema. It have two arguments, the first one is the **Query** we need to
+fetch (we will handle this part soon), in the second, we use the `props`
+property in the second parameter to compute the **Query** result `data` to a
+`loading` and a `store` properties.
+
+You can find all the available `graphql` options in the
+[**React Apollo** documentation](https://www.apollographql.com/docs/react/api/react-apollo.html)
+
+As you can see, our Enhancer need a **Query**, this a `.gql` (or `.graphql`)
+file that use the **GraphQL** syntax. In our case, it will look like:
```
Comment (JulienPradet) [inline docs/create-a-business-component.md]: I'd add a note that says how to build the query. Or a link to graphiql?
